cmd/main.go

```
@@ -2,45 +2,122 @@ package main
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
 	}
 
-	projectID := os.Getenv("GOOGLE_PROJECT_ID")
-	if projectID == "" {
-		log.Fatalf("GOOGLE_PROJECT_ID variable is not defined")
+	// Wait for interrupt.
+	<-ctx.Done()
+
+	if err := shutdownHandle.Call(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
 	}
+}
+
+// run executes the main program. It starts a server.Server inside
+// a goroutine and returns a configured shutdown.Handle for gracefully
+// shutting down the program.
+//
+// If the bootstrap of the server fails, run returns a non-nil error.
+// Otherwise the caller must use the shutdown.Handle to stop the server.
+func run(ctx context.Context, cancel context.CancelFunc) (*shutdown.Handle, error) {
+	configPath := flag.String("config", defaultConfigPath, "Path to the configuration file (for example --config .env)")
+
+	flag.Parse()
 
-	collectionID := os.Getenv("FIRESTORE_COLLECTION_ID")
-	if projectID == "" {
-		log.Fatalf("FIRESTORE_COLLECTION_ID variable is not defined")
+	config, err := readConfigFile(*configPath)
+	if err != nil {
+		return &shutdown.Handle{}, err
 	}
 
-	rs, err := firestore.NewReportService(context.Background(), projectID, collectionID)
+	rs, err := firestore.NewReportService(context.Background(), config.project, config.collection)
 	if err != nil {
-		log.Fatal(err)
+		return &shutdown.Handle{}, err
+	}
+
+	srv := server.New(config.addr, rs)
+
+	handle := shutdown.NewHandle(func() error {
+		if err := srv.Shutdown(ctx); err != nil && err != context.Canceled {
+			return err
+		}
+
+		if err := rs.Close(); err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	go func() {
+		if err := srv.Start(); err != nil && err != http.ErrServerClosed {
+			fmt.Fprintln(os.Stderr, err)
+			cancel()
+		}
+	}()
+
+	return handle, nil
+}
+
+type config struct {
+	addr       string
+	project    string
+	collection string
+}
+
+func defaultConfig() config {
+	return config{
+		addr: ":" + defaultPort,
+	}
+}
```

[suggestion] Maybe `defaultConfig` a bit misleading because it returns an invalid config, only the default port is set. I'd suggest to manage the port defaulting inside `readConfigFile` without relying on that function.